lec08-stability.Rmd

```diff
@@ -1,4 +1,576 @@
+    }
+
+})
+
+```
+
+INSERT GIF from school computer
+
+### Drawing qualitative phase portraits by hand
+
+Before we look at how to add more to phase portraits in R, let's pratice the process by hand to
+get familiar with it. Let's use the following two-dimensional system as an example. The two state
+variables are $x$ (position) and $v$ (velocity), and this is a model of a mass attached to a 
+spring. $dx/dt$ is the rate of change of position, and $dv/dt$ is the rate of change of velocity. 
+
+$$\frac{dx}{dt} = v$$
+
+$$\frac{dv}{dt} = -\omega^2 x$$
+
+We want to draw a **vector field** in two dimensions to show how the system will evolve in time.
+Our axes are $x$ and $v$. It doesn't matter which one is the vertical axis and which one is
+horizontal, but let's put $x$ on the x-axis and $v$ on the y-axis. These variables aren't 
+population sizes, so they can realistically be negative and non-integers. 
+
+Next, we choose some pairs of $x$ and $v$ and evaluate both $dx/dt$ and $dv/dt$, drawing an
+arrow at $(x,v)$ in the direction of the derivative. Let's also set $\omega = 1$ for convenience.
+Let's start with $(0,0)$: both derivatives are 0, so there's no arrow. For $(0,1)$, $dx/dt = 1$
+and $dv/dt = 0$, so we draw an arrow pointing horizontally in the $+x$ direction. Similarly for
+$(1,0)$. For $(1,1)$, we draw an arrow that goes $1$ unit in the $x$ direction and $-1$ unit in
+the $y$ direction. As we keep filling in this grid, we start to get a picture of how the system
+can behave for different starting points of $x$ and $v$. 
+
+This is analogous to plotting $dN/dt$ vs. $N$ and drawing arrows to the right where $dN/dt$ is
+positive and arrows to the left when it's negative, except that now we have to consider that 
+the rate of change is affecting both variables at the same time. 
+
+#### Challenge
+
+Define a function called `mass_spring` that uses the format required by `ode` to calculate 
+and return $dx/dt$ and $dv/dt$. We will use this function later to plot a phase portrait in R.
+Remember that the arguments for your function must be called `t`, `y`, and `parameters`. 
+
+```{r, eval=FALSE, echo=FALSE}
+# Challenge solution
+
+mass_spring <- function(t, y, parameters) {
+  # calculates dx/dt and dy/dt for a predator-prey model
+  
+  # t: time at which to evaluate derivatives
+  # y: vector of system variables (c(x, v))
+  # parameters: vector of model parameters (c(omega))
+    
+  # the state vector y has two elements because we have two variables
+  x <- y[1] # position
+  v <- y[2] # velocity
+  
+  omega <- parameters[1]
+    
+  # calculate rate of change
+  dx <- v
+  dv <- -omega^2 * x
+  
+  # return rate of change
+  return(list(c(dx, dv)))
+}
+```
+
+### Using `phaseR` to draw phase portraits in R
+
+`phaseR` is a package specifically for drawing phase portraits. It can automatically calculate 
+and plot things like trajectories and vector fields. 
+
+```{r}
+library(phaseR)
```

Comment:
You'll also need to do `devtools::use_package("phaseR")` for this package as well